lib/awful/widget/taglist.lua

```diff
@@ -149,9 +150,27 @@ taglist.filter = {}
 -- @beautiful beautiful.taglist_spacing
 -- @tparam[opt=0] number spacing The spacing between tags.
 
+--- The shape used for taglist widget container itself.
```

Comment:
I am not sure I like this. You are just adding random code to the taglist (but not the tasklist) that has no advantage over the user just wrapping her/his/it `taglist` in a `background` container. I see this the same way as adding `set_width/set_height/set_margin` to all widgets instead of using the idiomatic `margin` container and `forced_*` .

I understand your theme looks nicer with it, but I disagree it is a reason to extent the widgets. A future proof solution would be to add a `beautiful.main_wibar` (needs a better name) and use `beautiful.main_wibar or wibox.wibar {...}` in `rc.lua`. This way all those "but, but, it looks better!" changes could be implemented per theme instead of adding yet more options to the `taglist`/`tasklist`. It this point, I think I see the taglist/tasklist has in perpetual feature freeze.

Reply:
that will make theme not really reusable in configs with any custom widgets involved

Reply:
mb smth like `beautiful.taglist_base_widget` ?

Reply:
And adding unlimited options to taglist/tasklist *already* make it unmaintainable. Also, people could just modify the themes instead of `rc.lua`. In the long run, this is a **good** thing since we might have more themes usable with the default config instead of LCPZ/Copycat-killer `rc.lua` forks for every configs.

Visual code should go in the theme, window management logic in `rc.lua`. If we can find the right balance and have a nice separation of concern between the view and the logic, everybody wins.

Reply:
> beautiful.taglist_base_widget

It would not help, it will even make it worst. Think of use case such as the `tasklist` being used for both the wibar and in an alt+tab like popup. In that case adding too many options will make it impossible for module to reuse the tasklist without making sure to override every single style variable just to ensure it works.

Reply:
that approach makes harder to just switch themes with having single rc.lua

btw we `widget_template` in tasklist, having something like that there will be useful for more uniform API across default wibar widgets

Reply:
so, speaking about balance, i think the right one will be instead of providing only options in beautiful add more shortcuts for overriding standard widgets' contents

Reply:
but that's already a documented API in tasklist itself

Reply:
> but that's already a documented API in tasklist itself 

Maybe we are not talking about the same thing, can you provide a documentation link?

> Visual code should go in the theme

Yes, but I am thinking about higher level code (the whole wibar, notification and client decoration). Style options on the widget level are required to reuse them for different purpose (see the layoutlist in the other PR, it is a better example than the taglist). But IMO this is a longer term feature, not something for this PR. For now I am just against adding `widget_template` in the taglist beautiful namespace and wraping the taglist into a background for what I consider the wrong reasons. However, `taglist_spacing` is ok. While it contradict what I just said a bit, you are right that it is in line with what's already exposed to `beautiful`. Adding the background and widget_template, on the other hand will break things for many valid use cases.

Reply:
1) https://awesomewm.org/apidoc/classes/awful.widget.tasklist.html#awful.tasklist, second argument

2) i still think what mooving thing like a whole wibar to the theme will be not really reusable because wibar layout itself is not a matter of styling only but a matter of functionality, so this proposa means what any theme customization made on wibar will be skipped if user want to have custom wibar layout (ie wants to add one more custom widget to default wibar layout)

Reply:
btw, that's one more example why `beautiful.main_wibar` approach is bad (in the example it's called `beautiful.at_screen_connect(s)` but seems to be doing the same job as beautiful.main_wibar would do):

https://github.com/actionless/awesome_config/issues/24

Reply:
so that will be a painful decision, to make tasklist and taglist code more complicated, but we have to make to avoid even harder decisions in beautiful or leaving users to modify rc.lua for things which are obviousle part of the theme (like widgets' rounding)

Reply:
I really don't think doing changes to the `awful.widget.common` based widgets just to add more theme variables is a good idea. You are right that fixing it using higher level means is more complicated and will require long discussions and such. However doing the changes you propose here is a dead end. It will always be "but I only need one more" over and over again.

https://github.com/awesomeWM/awesome/issues/442 is "right" in the sense 99% of users `rc.lua` are either the default one + few changes or based on @lcpz templates. This hurt the project in the long run because people assume Awesome is inflexible while our very mission is to be flexible. At some point the issue have to be addressed one way or another. Burying our head in the sand and having the core developers adding more and more hardcoded things in those widgets to fit their current taste or the latest trends is not solving anything. It just proves those who think Awesome is inflexible right. If we have to modify core components to get anything done, then we failed somewhere.

The reason I pushed the "delegate" system was to put a definitive end on those options. What you propose here duplicates existing functionalities (the `wibox.container.background` container. You can say the `textclock`/`layoutbox`/`menu`/`systray` also lacks a shape, do you really think it's a good idea to add such options and theme variables to every single things? I admit I myself went a little too far with them and if I had implemented the delegates earlier I would not have added them to the taglist/tasklist. It was my mistake and my lazyness of not addressing the problem and mitigating the consequences.

Also the other argument that adding this options *breaks* the widget for alternate use cases is a big and definitive blocking issue. The `beautiful.tasklist_widget_template` **cannot** exist.

If you want this PR to ever be merged, only add the gtk theme and its helper function (and maybe more integration tests for them). Otherwise I don't see this going anywhere. Don't get me wrong, I am not against the concept of having a theme that follow the GTK palette. I think it is useful and worthy of inclusion. My issue are that the way it adds theme options is a no go even if there is no currently available viable alternatives.

Reply:
I see how your comment explains why implementing `beautiful.taglist_shape_*` is bad.

However i still can't undestand why `widget_template` can exist as argument to tasklist constructor but not as `beautiful.tasklist_widget_template`.

Reply:
Because if someone want to build an `alt+tab` widget using the tasklist like an example proposed in my `popup` PR does, then having a template potentially "damaged" by the theme variables forces any implementation to copy/paste the default delegate just in case the theme messed up the template in a way that renders such alt+tab widget useless.

Reply:
ok, that makes sense

i'll try to similify the theme and update the PR during the week